client-sniffer/src/main/java/org/elasticsearch/client/sniff/Sniffer.java

```diff
@@ -0,0 +1,216 @@
+        private final ScheduledExecutorService scheduledExecutorService;
+        private final AtomicBoolean running = new AtomicBoolean(false);
+        private ScheduledFuture<?> scheduledFuture;
+
+        private Task(HostsSniffer hostsSniffer, RestClient restClient, long sniffIntervalMillis, long sniffAfterFailureDelayMillis) {
+            this.hostsSniffer = hostsSniffer;
+            this.restClient = restClient;
+            this.sniffIntervalMillis = sniffIntervalMillis;
+            this.sniffAfterFailureDelayMillis = sniffAfterFailureDelayMillis;
+            this.scheduledExecutorService = Executors.newScheduledThreadPool(1);
+            scheduleNextRun(0);
+        }
+
+        synchronized void scheduleNextRun(long delayMillis) {
+            if (scheduledExecutorService.isShutdown() == false) {
+                try {
+                    if (scheduledFuture != null) {
+                        //regardless of when the next sniff is scheduled, cancel it and schedule a new one with updated delay
+                        this.scheduledFuture.cancel(false);
+                    }
+                    logger.debug("scheduling next sniff in " + delayMillis + " ms");
+                    this.scheduledFuture = this.scheduledExecutorService.schedule(this, delayMillis, TimeUnit.MILLISECONDS);
+                } catch(Throwable t) {
+                    logger.error("error while scheduling next sniffer task", t);
+                }
+            }
+        }
+
+        @Override
+        public void run() {
+            sniff(null, sniffIntervalMillis);
+        }
+
+        void sniffOnFailure(HttpHost failedHost) {
+            sniff(failedHost, sniffAfterFailureDelayMillis);
+        }
+
+        void sniff(HttpHost excludeHost, long nextSniffDelayMillis) {
+            if (running.compareAndSet(false, true)) {
+                try {
+                    List<HttpHost> sniffedNodes = hostsSniffer.sniffHosts();
+                    if (excludeHost != null) {
+                        sniffedNodes.remove(excludeHost);
+                    }
+                    logger.debug("sniffed nodes: " + sniffedNodes);
+                    this.restClient.setHosts(sniffedNodes.toArray(new HttpHost[sniffedNodes.size()]));
+                } catch (Throwable t) {
+                    logger.error("error while sniffing nodes", t);
+                } finally {
+                    scheduleNextRun(nextSniffDelayMillis);
+                    running.set(false);
+                }
+            }
+        }
+
+        synchronized void shutdown() {
+            scheduledExecutorService.shutdown();
+            try {
+                if (scheduledExecutorService.awaitTermination(1000, TimeUnit.MILLISECONDS)) {
+                    return;
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+            scheduledExecutorService.shutdownNow();
```

Comment:
about shutdown throwing InterruptedException. the method is called as part of close so we will have to deal with it ourselves I'm afraid anyways? Unless we require people to call shutdown separately. WDYT?